File: src/core/keys.ts
// Copyright 2026 Haridarman Kumaresan
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0; see LICENSE at repo root.

import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign as cryptoSign,
  verify as cryptoVerify,
} from "node:crypto";

/**
 * ed25519 keys are exchanged as base64url-encoded raw 32-byte values
 * (no padding). node:crypto only speaks DER, so we wrap/unwrap the raw
 * keys with the fixed ed25519 DER prefixes from RFC 8410.
 */
const PKCS8_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
const SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export interface Keypair {
  /** base64url raw 32-byte ed25519 public key */
  publicKey: string;
  /** base64url raw 32-byte ed25519 seed — keep private */
  secretKey: string;
}

export function toB64url(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64url");
}

export function fromB64url(s: string): Buffer {
  return Buffer.from(s, "base64url");
}

export function hexToBytes(hex: string): Buffer {
  return Buffer.from(hex, "hex");
}

export function utf8(s: string): Buffer {
  return Buffer.from(s, "utf8");
}

export function generateKeypair(): Keypair {
  const { publicKey, privateKey } = generateKeyPairSync("ed25519");
  const pub = publicKey.export({ type: "spki", format: "der" });
  const sec = privateKey.export({ type: "pkcs8", format: "der" });
  return {
    publicKey: toB64url(pub.subarray(SPKI_PREFIX.length)),
    secretKey: toB64url(sec.subarray(PKCS8_PREFIX.length)),
  };
}

function privateKeyObject(secretKey: string) {
  const seed = fromB64url(secretKey);
  if (seed.length !== 32) throw new RangeError("secretKey must be a base64url 32-byte ed25519 seed");
  return createPrivateKey({
    key: Buffer.concat([PKCS8_PREFIX, seed]),
    format: "der",
    type: "pkcs8",
  });
}

function publicKeyObject(publicKey: string) {
  const raw = fromB64url(publicKey);
  if (raw.length !== 32) throw new RangeError("publicKey must be a base64url 32-byte ed25519 key");
  return createPublicKey({
    key: Buffer.concat([SPKI_PREFIX, raw]),
    format: "der",
    type: "spki",
  });
}

export function publicKeyFromSecret(secretKey: string): string {
  const pem = privateKeyObject(secretKey).export({ type: "pkcs8", format: "pem" }) as string;
  const pub = createPublicKey(pem);
  return toB64url(pub.export({ type: "spki", format: "der" }).subarray(SPKI_PREFIX.length));
}

/** Sign raw bytes; returns base64url signature. */
export function signBytes(secretKey: string, data: Uint8Array): string {
  return toB64url(cryptoSign(null, data, privateKeyObject(secretKey)));
}

/**
 * Domain-separated signing. Every Countersign signature commits to a context
 * label so a signature minted for one envelope type can never be replayed as
 * another (intent vs. countersignature vs. email link). The signed bytes are
 * `context` + "\n" + `canonical`.
 */
export function signContext(secretKey: string, context: string, canonical: string): string {
  return signBytes(secretKey, utf8(`${context}\n${canonical}`));
}

/** Verify a domain-separated signature. Never throws on bad input. */
export function verifyContext(publicKey: string, context: string, canonical: string, signature: string): boolean {
  return verifyBytes(publicKey, utf8(`${context}\n${canonical}`), signature);
}

/** Verify a base64url signature over raw bytes. Never throws on bad input. */
export function verifyBytes(publicKey: string, data: Uint8Array, signature: string): boolean {
  try {
    return cryptoVerify(null, data, publicKeyObject(publicKey), fromB64url(signature));
  } catch {
    return false;
  }
}

/** Verify with raw byte inputs (used for Discord's hex-encoded headers). */
export function verifyRaw(publicKey: Uint8Array, data: Uint8Array, signature: Uint8Array): boolean {
  try {
    return cryptoVerify(
      null,
      data,
      createPublicKey({
        key: Buffer.concat([SPKI_PREFIX, Buffer.from(publicKey)]),
        format: "der",
        type: "spki",
      }),
      signature,
    );
  } catch {
    return false;
  }
}

File: clients/python/countersign/keys.py
# Copyright 2026 Haridarman Kumaresan
# SPDX-License-Identifier: Apache-2.0
"""
ed25519 keys and domain-separated signing — matches src/core/keys.ts.

Keys are exchanged as base64url of the raw 32-byte value (no padding): a 32-byte
seed for a secret, the 32-byte public key for a public key. The signed bytes for
every counter-sign signature are ``utf8(context + "\\n" + canonical)``.
"""
import base64

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat


def utf8(s: str) -> bytes:
    return s.encode("utf-8")


def b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def b64url_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def is_canonical_public_key(s) -> bool:
    """
    True iff ``s`` is the canonical base64url encoding of a raw 32-byte ed25519 key.
    base64url is not injective, so requiring canonicality (re-encoding the decoded
    bytes yields the identical string) gives every key exactly one string form —
    making exact-string comparison of keys sound.
    """
    if not isinstance(s, str):
        return False
    try:
        raw = b64url_decode(s)
    except Exception:
        return False
    return len(raw) == 32 and b64url_encode(raw) == s


def _private_key(secret: str) -> Ed25519PrivateKey:
    seed = b64url_decode(secret)
    if len(seed) != 32:
        raise ValueError("secret must be a base64url 32-byte ed25519 seed")
    return Ed25519PrivateKey.from_private_bytes(seed)


def public_key_from_secret(secret: str) -> str:
    pub = _private_key(secret).public_key()
    return b64url_encode(pub.public_bytes(Encoding.Raw, PublicFormat.Raw))


def sign_bytes(secret: str, data: bytes) -> str:
    return b64url_encode(_private_key(secret).sign(data))


def sign_context(secret: str, context: str, canonical: str) -> str:
    """Domain-separated signing: sign over ``utf8(context + "\\n" + canonical)``."""
    return sign_bytes(secret, utf8(f"{context}\n{canonical}"))


def verify_bytes(public: str, data: bytes, signature: str) -> bool:
    """Verify a base64url signature over raw bytes. Never raises on bad input."""
    try:
        Ed25519PublicKey.from_public_bytes(b64url_decode(public)).verify(
            b64url_decode(signature), data
        )
        return True
    except Exception:
        return False


def verify_context(public: str, context: str, canonical: str, signature: str) -> bool:
    return verify_bytes(public, utf8(f"{context}\n{canonical}"), signature)
